File: lexo-backend/api/practice.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from core.database import get_db
from game.logic import generate_letter_pool, calculate_score, has_letters_in_pool
from game.word_list import is_word_valid
from typing import List, Dict, Any
from pydantic import BaseModel
import uuid
import time

router = APIRouter()

practice_sessions: Dict[str, Dict[str, Any]] = {}

class PracticeSession(BaseModel):
    session_id: str
    letter_pool: List[str]
    score: int
    words_found: List[str]
    time_started: float
    duration: int

class WordSubmission(BaseModel):
    word: str

class PracticeStartRequest(BaseModel):
    duration: int = 300

@router.post("/practice/start")
def start_practice_session(request: PracticeStartRequest, db: Session = Depends(get_db)):
    session_id = str(uuid.uuid4())
    
    letter_pool = generate_letter_pool(12)
    
    session = {
        "session_id": session_id,
        "letter_pool": letter_pool,
        "score": 0,
        "words_found": [],
        "used_words": set(),
        "time_started": time.time(),
        "duration": request.duration,
        "is_active": True
    }
    
    practice_sessions[session_id] = session
    
    return {
        "session_id": session_id,
        "letter_pool": letter_pool,
        "score": 0,
        "words_found": [],
        "time_remaining": request.duration
    }

@router.post("/practice/{session_id}/submit")
def submit_word(session_id: str, submission: WordSubmission, db: Session = Depends(get_db)):
    if session_id not in practice_sessions:
        raise HTTPException(status_code=404, detail="Practice session not found")
    
    session = practice_sessions[session_id]
    
    elapsed_time = time.time() - session["time_started"]
    if elapsed_time >= session["duration"]:
        session["is_active"] = False
        raise HTTPException(status_code=400, detail="Practice session has ended")
    
    if not session["is_active"]:
        raise HTTPException(status_code=400, detail="Practice session is not active")
    
    word = submission.word.lower().strip()
    
    if word in session["used_words"]:
        return {
            "success": False,
            "message": f'"{word}" has already been used in this session.',
            "score": session["score"],
            "letter_pool": session["letter_pool"],
            "time_remaining": max(0, session["duration"] - elapsed_time)
        }
    
    if not has_letters_in_pool(word, session["letter_pool"]):
        return {
            "success": False,
            "message": f'Not enough letters in the pool for "{word}".',
            "score": session["score"],
            "letter_pool": session["letter_pool"],
            "time_remaining": max(0, session["duration"] - elapsed_time)
        }
    
    if not is_word_valid(word):
        return {
            "success": False,
            "message": f'"{word}" is not a valid Turkish word.',
            "score": session["score"],
            "letter_pool": session["letter_pool"],
            "time_remaining": max(0, session["duration"] - elapsed_time)
        }
    
    word_score = calculate_score(word)
    session["score"] += word_score
    session["words_found"].append(word)
    session["used_words"].add(word)
    
    temp_pool = session["letter_pool"].copy()
    for letter in word:
        temp_pool.remove(letter)
    
    new_letters = generate_letter_pool(len(word))
    session["letter_pool"] = temp_pool + new_letters
    
    time_remaining = max(0, session["duration"] - elapsed_time)
    
    return {
        "success": True,
        "message": f'"{word}" is valid! +{word_score} points',
        "word": word,
        "word_score": word_score,
        "total_score": session["score"],
        "letter_pool": session["letter_pool"],
        "words_found": session["words_found"],
        "time_remaining": time_remaining
    }

@router.get("/practice/{session_id}/status")
def get_practice_status(session_id: str, db: Session = Depends(get_db)):
    if session_id not in practice_sessions:
        raise HTTPException(status_code=404, detail="Practice session not found")
    
    session = practice_sessions[session_id]
    elapsed_time = time.time() - session["time_started"]
    time_remaining = max(0, session["duration"] - elapsed_time)
    
    if elapsed_time >= session["duration"]:
        session["is_active"] = False
    
    return {
        "session_id": session_id,
        "letter_pool": session["letter_pool"],
        "score": session["score"],
        "words_found": session["words_found"],
        "time_remaining": time_remaining,
        "is_active": session["is_active"]
    }

@router.post("/practice/{session_id}/end")
def end_practice_session(session_id: str, db: Session = Depends(get_db)):
    if session_id not in practice_sessions:
        raise HTTPException(status_code=404, detail="Practice session not found")
    
    session = practice_sessions[session_id]
    session["is_active"] = False
    
    elapsed_time = time.time() - session["time_started"]
    actual_duration = min(elapsed_time, session["duration"])
    
    results = {
        "session_id": session_id,
        "final_score": session["score"],
        "words_found": session["words_found"],
        "total_words": len(session["words_found"]),
        "duration": actual_duration,
        "words_per_minute": len(session["words_found"]) / (actual_duration / 60) if actual_duration > 0 else 0
    }
    
    return results

File: lexo-frontend/src/api/practice.ts
const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || "http://localhost:8000/api";

export interface PracticeSession {
  session_id: string;
  letter_pool: string[];
  score: number;
  words_found: string[];
  time_remaining: number;
}

export interface PracticeStartRequest {
  duration?: number;
}

export interface WordSubmissionRequest {
  word: string;
}

export interface PracticeResults {
  session_id: string;
  final_score: number;
  words_found: string[];
  total_words: number;
  duration: number;
  words_per_minute: number;
}

export const startPracticeSession = async (duration: number = 300): Promise<PracticeSession> => {
  const token = sessionStorage.getItem('access_token');
  const response = await fetch(`${API_BASE_URL}/practice/start`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` }),
    },
    body: JSON.stringify({ duration }),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to start practice session');
  }

  return response.json();
};

export const submitPracticeWord = async (sessionId: string, word: string) => {
  const token = sessionStorage.getItem('access_token');
  const response = await fetch(`${API_BASE_URL}/practice/${sessionId}/submit`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` }),
    },
    body: JSON.stringify({ word }),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to submit word');
  }

  return response.json();
};

export const getPracticeStatus = async (sessionId: string) => {
  const token = localStorage.getItem('access_token');
  const response = await fetch(`${API_BASE_URL}/practice/${sessionId}/status`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` }),
    },
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to get practice status');
  }

  return response.json();
};

export const endPracticeSession = async (sessionId: string): Promise<PracticeResults> => {
  const token = localStorage.getItem('access_token');
  const response = await fetch(`${API_BASE_URL}/practice/${sessionId}/end`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token && { Authorization: `Bearer ${token}` }),
    },
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Failed to end practice session');
  }

  return response.json();
};
